cmd/generator/main.go

```diff
@@ -213,39 +234,73 @@ func getReleasePlugin(ctx context.Context, client *github.Client, repositoryName
 			return nil, errors.Wrapf(err, "failed to read gzipped plugin bundle for release %s", releaseName)
 		}
 
-		bundleReader := tar.NewReader(gzBundleReader)
-		for {
-			hdr, err := bundleReader.Next()
-			if err == io.EOF {
-				break
-			}
-			if err != nil {
-				return nil, errors.Wrapf(err, "failed to read plugin bundle for release %s", releaseName)
-			}
+		bundleData, err := ioutil.ReadAll(gzBundleReader)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to read plugin bundle for release %s", releaseName)
+		}
+
+		manifestData, err := getFromTarFile(tar.NewReader(bytes.NewReader(bundleData)), "plugin.json")
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to read manifest from plugin bundle for release %s", releaseName)
+		}
+		plugin.Manifest = mattermostModel.ManifestFromJson(bytes.NewReader(manifestData))
 
-			if path.Base(hdr.Name) != "plugin.json" {
-				continue
+		if plugin.Manifest.IconPath != "" {
+			iconData, err := getFromTarFile(tar.NewReader(bytes.NewReader(bundleData)), plugin.Manifest.IconPath)
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to read icon data from plugin bundle for release %s", releaseName)
 			}
-			manifest := mattermostModel.ManifestFromJson(bundleReader)
 
-			plugin.Manifest = manifest
-			break
+			logger.Debugf("using icon specified in manifest as %s", plugin.Manifest.IconPath)
+			plugin.IconData = fmt.Sprintf("data:image/svg+xml;base64,%s", base64.StdEncoding.EncodeToString(iconData))
 		}
 	} else {
 		logger.Debugf("skipping download since found existing plugin")
 	}
 
+	if plugin.Manifest == nil {
+		return nil, fmt.Errorf("failed to find plugin manifest for release %s", releaseName)
+	}
```

Comment:
Should this check be before we try to get the `IconPath` in:

``if plugin.Manifest.IconPath != "" {``

line 248

Reply:
We write to `plugin.Manifest` just before that:
https://github.com/mattermost/mattermost-marketplace/blob/9a9787e3bbad25cb45ca6649e3facf94ccdd903b/cmd/generator/main.go#L246

but I suppose this could still return `nil`. I'll add a check.